File: Experience/JavaScript/js/LevenshteinDistance.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const levDist = (str1, str2) => {
    const m = str1.length;
    const n = str2.length;
    if (m == 0) {
        return m;
    }
    if (n == 0) {
        return m;
    }
    const d = [];
    for (let i = 0; i <= m; i++) {
        let row = new Array(n + 1).fill(0);
        d.push(row);
    }
    for (let i = 1; i <= m; i++) {
        d[i][0] = i;
    }
    for (let j = 1; j <= n; j++) {
        d[0][j] = j;
    }
    for (let j = 1; j <= n; j++) {
        for (let i = 1; i <= m; i++) {
            const substitutionCost = str1.charAt(i - 1) === str2.charAt(j - 1) ? 0 : 1;
            d[i][j] = Math.min(d[i - 1][j] + 1, d[i][j - 1] + 1, d[i - 1][j - 1] + substitutionCost);
        }
    }
    return d[m][n];
};
const compare_distances = (words, start, end) => {
    console.log(`Thread dispatched for ${start}-${end}`);
    let res = [];
    for (let i = start; i < end; i++) {
        for (let j = i + 1; j < words.length; j++) {
            let distance = levDist(words[i], words[j]);
            res.push({ i, j, distance });
        }
    }
    console.log(`Thread finished for ${start}-${end}`);
    return res;
};
const main = () => __awaiter(void 0, void 0, void 0, function* () {
    const args = process.argv;
    if (args.length < 4)
        throw new Error("Please provide arguments: <file_path> <num_threads>");
    let file_path = args[2];
    let num_threads = parseInt(args[3]);
    const strings = fs
        .readFileSync(file_path)
        .toString()
        .split("\n")
        .filter((str) => str !== ""); // filter out any falsy values (like empty line)
    let chunk_size = Math.floor((strings.length + num_threads - 1) / num_threads);
    let promises = [];
    for (let i = 0; i < num_threads; i++) {
        promises.push(new Promise((resolve, _) => {
            let start = i * chunk_size;
            let end = Math.min((i + 1) * chunk_size, strings.length);
            let distances = compare_distances(strings, start, end);
            resolve(distances);
        }));
    }
    let responses = yield Promise.all(promises);
    let comparisons = responses.flat();
    let bestRes = comparisons.reduce((prev, curr) => curr.distance < prev.distance ? curr : prev);
    console.log(`Most similar strings are: \n\t${strings[bestRes.i]}\n\t${strings[bestRes.j]}\nWith distance of ${bestRes.distance}`);
});
main();

File: Experience/C++/LevenshteinDistance.cpp
#include <algorithm>
#include <iostream>
#include <fstream>
#include <string.h>
#include <thread>
#include <vector>
#include <mutex>
#include <queue>

using namespace std;


int levenshtein_distance(const string& str1, const string& str2) {
	int m = str1.length(); // rows of matrix = length of str1 
	int n = str2.length(); // columns of matrix = length of str2  

	// Create a matrix to store results of subproblems (m+1)*(n+1) size  
	vector<vector<int>> dp((n + 1), vector<int>(m + 1, 0));

	// Initialize first row and column as 0  
	for (int i = 1; i <= m; i++) { dp[i][0] = i; }
	for (int j = 1; j <= n; j++) { dp[0][j] = j; }

	for (int j = 1; j <= n; j++) {
		for (int i = 1; i <= m; i++) {
			const int substitutionCost = (str1[i - 1] == str2[j - 1]) ? 0 : 1;

			dp[i][j] = min(dp[i - 1][j] + 1, min(dp[i][j - 1] + 1, dp[i - 1][j - 1] + substitutionCost));
		}
	}

	return dp[m][n];
}

void compare_distances(const vector<string>& words, size_t start, size_t end,
	queue<vector<int>>& results, mutex& mtx) {
	cout << "Thread dispatched for " << start << "-" << end << endl;

	for (size_t i = start; i < end; i++) {
		for (size_t j = i + 1; j < words.size(); j++) {
			int distance = levenshtein_distance(words[i], words[j]);
			lock_guard<mutex> lock(mtx);
			if (i != j) { 
				results.push({ (int) i, (int) j, distance }); };
		}
	}

	cout << "Thread finished for " << start << "-" << end << endl;
}
int main(int argc, char* argv[]) {
	// Get command line arguments
	if (argc < 3) {
		cerr << "Please give path to text file and number of threads" << endl;
		return 1;
	}

	string file_path(argv[1]);

	int num_threads = stoi(argv[2]);

	// Read the input strings from the file
	ifstream file(file_path);
	if (!file.is_open()) {
		cerr << "Unable to open file" << endl;
		return 1;
	}

	vector<string> words;
	string line;
	while (getline(file, line)) {
		words.push_back(line);
	}
	file.close();

	// Create a queue for communication between the master thread and worker threads
	queue<vector<int>> results;
	mutex mtx;

	// Calculate chunk size
	size_t chunk_size = (words.size() + num_threads - 1) / num_threads;

	// Dispatch worker threads
	vector<thread> thread_handles;
	for (int i = 0; i < num_threads; ++i) {
		auto start = i * chunk_size;
		auto end = min((i + 1) * chunk_size, words.size());
		auto sub_words = vector<string>(words.begin() + start, words.begin() + end);

		thread_handles.push_back(thread(compare_distances, words, start, end,
			ref(results), ref(mtx)));
	}

	// Wait for all threads to finish
	for (auto& handle : thread_handles) {
		handle.join();
	}

	// Collect the results and find the smallest distance
	int min_distance = INT32_MAX;
	vector<size_t> min_indices;
	for (size_t i = 0; i < words.size() * (words.size() - 1) / 2; ++i) {
		mtx.lock();
		if (!results.empty()) {
			auto res = results.front();
			size_t i = res[0];
			size_t j = res[1];
			size_t distance = res[2];
			results.pop();
			if (distance < min_distance) {
				min_distance = distance;
				min_indices = { i, j };
			}
		}
		mtx.unlock();
	}

	// Print the most similar strings
	cout << "Most similar string are: \n\t" << words[min_indices[0]] << "\n\t"
		<< words[min_indices[1]] << "\nWith distance of " << min_distance << endl;

	

	return 0;
}